delegated_credentials.go

```diff
@@ -0,0 +1,295 @@
+		if version == 0 {
+			return nil, errors.New("tls: Only TLS 1.2 or 1.3 are supported")
+		}
+		if config.timeFunc == nil {
+			config.timeFunc = time.Now
+		}
+
+		config.scheme = selectedScheme
+		config.version = version
+		err := fetchCredentialFromCache(config)
+		if err != nil {
+			return nil, err
+		}
+
+		config.Cert.DelegatedCredential = cachedCredentials[selectedScheme].credential
+		config.Cert.PrivateKey = cachedCredentials[selectedScheme].privateKey
+		return config.Cert, nil
+	}
+}
+
+// Checks if a credential is already stored in memory. If not or the credential
+// is already expired a new credential is created.
+func fetchCredentialFromCache(config *CredentialsConfig) (err error) {
+	if cachedCredentials == nil {
+		cachedCredentials = make(map[SignatureScheme]*cachedCredential)
+	}
+	if cachedCredentials[config.scheme] == nil {
+		cachedCredentials[config.scheme], err = newCredential(config)
+	} else {
+		if config.timeFunc().After(cachedCredentials[config.scheme].renewAfter) {
+			go func() {
+				cachedCredentials[config.scheme], err = newCredential(config)
+				if config.renewChan != nil {
+					config.renewChan <- struct{}{}
+				}
+			}()
+		}
+	}
+	return
+}
+
+func newCredential(config *CredentialsConfig) (*cachedCredential, error) {
+	credential, privateKey, renewAfter, err := createDelegatedCredential(config)
+	credentialBytes, err := credential.marshalAndSign(config.Cert, config.scheme, config.version)
+	if err != nil {
+		return nil, fmt.Errorf("tls: creating Delegated Credential failed (%s)", err)
+	}
+	return &cachedCredential{
+		credential: credentialBytes,
+		validTime:  credential.ValidTime,
+		privateKey: privateKey,
+		renewAfter: renewAfter,
+	}, nil
+}
+
+func selectVersion(versions []uint16) uint16 {
+	for _, version := range versions {
+		if version == VersionTLS13 {
+			return VersionTLS13
+		} else if version == VersionTLS12 {
+			return VersionTLS12
+		}
+	}
+	return 0
+}
+
+// Selects signature scheme based on the client's advertised schemes and the cert's capabilities
+func selectSignatureScheme(signatureSchemes []SignatureScheme, cert *x509.Certificate) SignatureScheme {
+	for _, scheme := range signatureSchemes {
+		if cert.PublicKeyAlgorithm == x509.ECDSA && scheme == ECDSAWithP256AndSHA256 {
+			return ECDSAWithP256AndSHA256
+		} else if cert.PublicKeyAlgorithm == x509.RSA {
+			if scheme == PSSWithSHA256 {
+				return PSSWithSHA256
+			} else if scheme == PKCS1WithSHA256 {
+				return PKCS1WithSHA256
+			}
+		}
+	}
+	return 0
+}
+
+// Creates new Delegated Credential. The type of the credential is decided based on the selected
+// SignatureScheme to ensure client's support.
+func createDelegatedCredential(config *CredentialsConfig) (credential DelegatedCredential, privateKey crypto.PrivateKey, renewAfter time.Time, err error) {
+	validTill := config.timeFunc().Add(config.Validity)
+	relativeToCert := validTill.Sub(config.Cert.Leaf.NotBefore)
```

Comment:
The computation of the ValidTime is a rather complex dance. Is there a simpler way to compute it that is more readable?